README.md
# RactorPool

TODO: Delete this and the text below, and describe your gem

Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/ractor_pool`. To experiment with that code, run `bin/console` for an interactive prompt.

## Installation

TODO: Replace `UPDATE_WITH_YOUR_GEM_NAME_PRIOR_TO_RELEASE_TO_RUBYGEMS_ORG` with your gem name right after releasing it to RubyGems.org. Please do not do it earlier due to security reasons. Alternatively, replace this section with instructions to install your gem from git if you don't plan to release to RubyGems.org.

Install the gem and add to the application's Gemfile by executing:

    $ bundle add UPDATE_WITH_YOUR_GEM_NAME_PRIOR_TO_RELEASE_TO_RUBYGEMS_ORG

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install UPDATE_WITH_YOUR_GEM_NAME_PRIOR_TO_RELEASE_TO_RUBYGEMS_ORG

## Usage

TODO: Write usage instructions here

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/ractor_pool. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/ractor_pool/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the RactorPool project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/ractor_pool/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2023-03-18

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true
require 'pry'
require 'ractor_pool'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/ractor_pool_spec.rb
# frozen_string_literal: true

class Fibonacci
  def call(n) = [n, fib(n)]
  def fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
end

module FibWork
  def self.call(n) = [n, fib(n)]
  def self.fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
end

RSpec.describe RactorPool do
  it 'has a version number' do
    expect(RactorPool::VERSION).not_to be nil
  end

end


spec/ractor_pool/single_use_spec.rb
# frozen_string_literal: true

require 'spec_helper'
require 'ractor_pool_spec'
require 'ractor_pool/single_use'

RSpec.describe RactorPool::SingleUse do
  subject(:pool) { described_class.call(inputs, behavior, options) }

  let(:options) { {} }
  let(:expected_ractor_count) { (options[:worker_count] || 1) + 1 + 1 } # 1 for the main ractor & 1 for the queue

  before { expect(Ractor.count).to eq(1) }
  after { expect(Ractor.count).to eq(1) }

  context 'without a pool' do
    it { expect(Ractor.count).to eq(1) }
  end

  context 'when behavior is a class' do
    let(:behavior) { Fibonacci.new }

    context 'with a single input' do
      let(:inputs) { 5 }
      let(:expected_results) { [[5, 8]] }

      context 'when the pool is created' do
        after { pool.results; sleep(0.1) }

        it { expect(pool).to be_a(described_class) }

        it 'it creates the expected number of Ractors' do
          pool
          expect(Ractor.count).to eq(expected_ractor_count)
        end
      end

      it 'returns the expected result' do
        expect(pool.results).to eq(expected_results)
        sleep(0.1)
      end
    end

    context 'with multiple inputs' do
      let(:inputs) { [4, 5, 6, 7] }
      let(:options) { { worker_count: 3 } }
      let(:expected_results) { [[4, 5], [5, 8], [6, 13], [7, 21]] }

      context 'when the pool is created' do
        after { pool.results; sleep(0.1) }

        it 'it creates an additional ractor for each input' do
          options[:worker_count] = 3
          pool
          expect(Ractor.count).to eq(expected_ractor_count)
        end
      end

      it 'returns the expected result' do
        expect(pool.results).to match_array(expected_results)
        sleep(0.1)
      end
    end
  end

  context 'when behavior is a module' do
    let(:behavior) { FibWork }

    context 'with a single input' do
      let(:inputs) { 5 }
      let(:expected_results) { [[5, 8]] }

      context 'when the pool is created' do
        after { pool.results; sleep(0.1) }

        it { expect(pool).to be_a(described_class) }

        it 'it creates an additional ractor for each input' do
          pool
          expect(Ractor.count).to eq(expected_ractor_count)
        end
      end

      context 'after the results are returned' do
        before do
          pool.results
          sleep(0.1) # Garbage collection needs time to run
        end

        it 'the Ractor count is reduced' do
          expect(Ractor.count).to eq(1) # 1 for the main ractor
        end
      end

      it 'returns the expected result' do
        expect(pool.results).to eq(expected_results)
        sleep(0.1)
      end
    end

    context 'with multiple inputs' do
      let(:inputs) { [5, 6, 7] }
      let(:expected_results) { [[5, 8], [6, 13], [7, 21]] }

      context 'when the pool is created' do
        after { pool.results; sleep(0.1) }

        it 'it creates an additional ractor for each input' do
          pool
          expect(Ractor.count).to eq(expected_ractor_count)
        end
      end

      context 'after the results are returned' do
        before do
          pool.results
          sleep(0.1) # Garbage collection needs time to run
        end

        it 'the Ractor count is reduced' do
          expect(Ractor.count).to eq(1) # 1 for the main ractor
        end
      end

      it 'returns the expected result' do
        expect(pool.results).to match_array(expected_results)
        sleep(0.1)
      end
    end

    context 'with malformed inputs' do
      let(:inputs) { [5, '6', 7] }
      let(:expected_results) { [[5, 8], [7, 21]] }

      context 'when the pool is created' do
        after { pool.results; sleep(0.1) }

        it 'it creates an additional ractor for each input' do
          pool
          expect(Ractor.count).to eq(expected_ractor_count)
        end
      end

      context 'after the results are returned' do
        before do
          pool.results
          sleep(0.1) # Garbage collection needs time to run
        end

        it 'the Ractor count is reduced' do
          expect(Ractor.count).to eq(1) # 1 for the main ractor
        end
      end

      it 'returns the expected result' do
        expect(pool.results).to match_array(expected_results)
        sleep(0.1)
      end
    end
  end

  xcontext 'when behavior is a Proc' do
    let(:inputs) { 5 }
    let(:behavior) { ->(n) { [n, n * 2] } }
    let(:expected_results) { [[5, 8]] }

    it 'returns the expected result' do
      expect(pool.results).to match_array(expected_results)
      sleep(0.1)
    end
  end
end

# # Ractor.count # 1
# # pool = Ractors::ForkedPool.new(5, Fibonacci.new)
# # Ractor.count # 2
# # results = pool.results
# # Ractor.count # 1

# # messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# # Ractor.count # 1
# # pool = Ractors::ForkedPool.new(messages, Fibonacci.new)
# # Ractor.count # 16
# # results = pool.results
# # Ractor.count # 1

# # module FibWork
# #   def self.call(n) = [n, fib(n)]
# #   def self.fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
# # end

# # Ractor.count # 1
# # pool = Ractors::ForkedPool.new(5, FibWork)
# # Ractor.count # 2
# # results = pool.results
# # Ractor.count # 1
# # results

# # messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# # Ractor.count # 1
# # pool = Ractors::ForkedPool.new(messages, FibWork)
# # Ractor.count # 16
# # results = pool.results
# # Ractor.count # 1
# # results


lib/ractor_pool/forked_join.rb
# frozen_string_literal: true

module RactorPool
  # using the ForkedJoin pattern - https://en.wikipedia.org/wiki/Fork%E2%80%93join_model
  # this creates a worker pool (of Ractors) as large as the number of input messages
  # and then collects the results upon completion - at which point the Ractors are
  # removed from the worker pool and available to be garbage collected
  # note: the injected behavior must respond to #call(input)
  # usage: define a behavior that responds to #call(input)
  # class Fibonacci
  #   def call(n) = [n, fib(n)]
  #   def fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
  # end
  # and pass it to the pool
  # pool = Ractors::ForkedPool.new(5, Fibonacci.new)
  # do other stuff
  # results = pool.results # this will block until all workers are done
  # Ractor.count # 1 (main ractor / thread)
  # more examples in examples and in spec/ractor_pool/forked_join_spec.rb
  class ForkedJoin
    attr_reader :behavior, :messages, :workers

    def self.call(inputs, behavior) = new(inputs, behavior)

    def initialize(inputs, behavior)
      @behavior = behavior
      @messages = Array(inputs)
      @workers = messages.map { |message| spawn_ractor(message) }
    end

    # removes ractors when done and GC removes
    def results
      results = []
      until workers.empty?
        begin
          worker, result = Ractor.select(*workers)
          results << result
        rescue Ractor::RemoteError => e
          worker = e.ractor
          puts e.inspect # log error
        ensure
          workers.delete(worker)
        end
      end
      results
    end

    private

    def spawn_ractor(message)
      ractor = Ractor.new do
        number, calculation = Ractor.receive
        Ractor.yield calculation.call(number)
      end
      ractor.send([message, behavior])
    end
  end
end

# class Fibonacci
#   def call(n) = [n, fib(n)]
#   def fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
# end

# Ractor.count # 1
# pool = RactorPool::ForkedPool.new(5, Fibonacci.new)
# Ractor.count # 2
# results = pool.results
# Ractor.count # 1

# messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# Ractor.count # 1
# pool = RactorPool::ForkedPool.new(messages, Fibonacci.new)
# Ractor.count # 16
# results = pool.results
# Ractor.count # 1


# module FibWork
#   def self.call(n) = [n, fib(n)]
#   def self.fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
# end

# Ractor.count # 1
# pool = RactorPool::ForkedPool.new(5, FibWork)
# Ractor.count # 2
# results = pool.results
# Ractor.count # 1
# results

# messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# Ractor.count # 1
# pool = RactorPool::ForkedPool.new(messages, FibWork)
# Ractor.count # 16
# results = pool.results
# Ractor.count # 1
# results


lib/ractor_pool/single_use.rb
# frozen_string_literal: true

module RactorPool
  # static sized pool with supervision
  # use multi_use at own risk - be sure to collect results before submitting more messages
  class SingleUse
    attr_reader :behavior, :inputs, :input_queue, :multi_use, :worker_count, :workers_pool

    def self.call(input, behavior, options = {})
      new(input, behavior, options)
    end

    def initialize(input, behavior, options = {})
      @worker_count = options[:worker_count] || 1
      @multi_use = (options[:multi_use] == true)

      @input_queue = create_input_queue
      @workers_pool = create_worker_pool

      @inputs = Array(input)
      @behavior = behavior

      inputs.each { |message| input_queue.send([message, behavior]) }
    end

    def results
      results = collect_results
      terminate
      results
    end

    private

    def spawn_worker
      Ractor.new(input_queue) do |queue|
        # pull work when available - terminate upon receiving nil
        until (message = queue.take).nil?
          value, comute = message
          Ractor.yield(comute.call(value))
        end
      end
    end

    def collect_results
      inputs.each_with_object([]) do |_i, answers|
        _worker, answer = Ractor.select(*workers_pool)
        answers << answer
      rescue Ractor::RemoteError => e
        dead_worker = e.ractor
        workers_pool.delete(dead_worker)
        workers_pool << spawn_worker
        puts e.inspect
      end
    end

    def create_input_queue
      Ractor.new do
        until (message = Ractor.receive) == :close
          Ractor.yield(message)
        end
      end
    end

    def terminate
      # kill all workers
      # worker_count.times { input_queue << nil }
      worker_count.times { input_queue.send(nil) }
      # kill input queue
      # input_queue << :close
      input_queue.send(:close)
      p 'terminated'
    end

    def create_worker_pool = (1..worker_count).map { |_i| spawn_worker }
  end
end

# Ractor.count # 1
# pool = RactorPool::SupervisedWorkers.call(5, Fibonacci.new)
# Ractor.count # 3 - 1 worker + 1 input queue + 1 main
# results = pool.results
# pp results
# Ractor.count # 1

# options = { worker_count: 6 }
# messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# Ractor.count # 1
# pool = RactorPool::SupervisedWorkers.call(messages, Fibonacci.new, options)
# Ractor.count # 8 - 6 workers + 1 input queue + 1 main
# results = pool.results
# pp results
# Ractor.count # 1

# options = { worker_count: 6, multi_use: true }
# messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# Ractor.count # 1
# pool = RactorPool::SupervisedWorkers.call(messages, Fibonacci.new, options)
# Ractor.count # 8 - 6 workers + 1 input queue + 1 main
# results1 = pool.results
# pp results1
# Ractor.count # 8 - 6 workers + 1 input queue + 1 main

# messages = [0, 1, 2, 3]
# pool.submit(messages, Fibonacci.new)
# Ractor.count # 8 - 6 workers + 1 input queue + 1 main
# results2 = pool.results
# restuls2
# Ractor.count # 8 - 6 workers + 1 input queue + 1 main
# pool.terminate

# Ractor.count # 1 - main

# # module FibWork
# #   def self.call(n) = [n, fib(n)]
# #   def self.fib(n) = n < 2 ? 1 : fib(n-2) + fib(n-1)
# # end

# # Ractor.count # 1
# # pool = RactorPool::ForkedPool.new(5, FibWork)
# # Ractor.count # 2
# # results = pool.results
# # Ractor.count # 1
# # results

# # messages = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
# # Ractor.count # 1
# # pool = RactorPool::ForkedPool.new(messages, FibWork)
# # Ractor.count # 16
# # results = pool.results
# # Ractor.count # 1
# # results
